prov/util/src/util_wait.c

```diff
@@ -142,6 +142,68 @@ int fi_wait_init(struct util_fabric *fabric, struct fi_wait_attr *attr,
 	return 0;
 }
 
+int ofi_wait_fd_match(struct dlist_entry *item, const void *arg)
+{
+	struct ofi_wait_fd_entry *fd_entry;
+
+	fd_entry = container_of(item, struct ofi_wait_fd_entry, entry);
+	return fd_entry->fd == *(int *)arg;
+}
+
+int ofi_wait_fd_del(struct util_wait *wait, int fd)
+{
+	struct ofi_wait_fd_entry *fd_entry;
+	struct dlist_entry *entry;
+	struct util_wait_fd *wait_fd = container_of(wait, struct util_wait_fd,
+						    util_wait);
+
+	fastlock_acquire(&wait_fd->lock);
+	entry = dlist_remove_first_match(&wait_fd->fd_list, ofi_wait_fd_match,
+					 &fd);
+	fastlock_release(&wait_fd->lock);
+	if (!entry) {
+		FI_WARN(wait->prov, FI_LOG_FABRIC,
+			"Given fd not found in wait list\n");
+		return -FI_EINVAL;
+	}
+
+	fd_entry = container_of(entry, struct ofi_wait_fd_entry, entry);
+	fi_epoll_del(wait_fd->epoll_fd, fd_entry->fd);
+	free(fd_entry);
+	return 0;
+}
+
+int ofi_wait_fd_add(struct util_wait *wait, int fd, ofi_wait_fd_try_func try,
+		    void *arg, void *context)
+{
+	struct ofi_wait_fd_entry *fd_entry;
+	struct util_wait_fd *wait_fd = container_of(wait, struct util_wait_fd,
+						    util_wait);
+	int ret = 0;
+
+	fastlock_acquire(&wait_fd->lock);
+	if (dlist_find_first_match(&wait_fd->fd_list, ofi_wait_fd_match, &fd))
+		goto out;
+
+	ret = fi_epoll_add(wait_fd->epoll_fd, fd, context);
+	if (ret)
+		goto out;
```

Comment:
Add an FI_WARN on error here.  We should be better about adding error messages on error paths to assist with future debugging efforts.